CRM/Contribute/BAO/Contribution.php

```diff
@@ -3824,6 +3824,37 @@ public static function validateFinancialType($financialTypeId, $relationName = '
     return FALSE;
   }
 
+  /**
+   * Function to process additional payment for partial and refund contributions. This function
+   * is called from backend 'Record Payment' form and also when a partial payment is completed
+   * from the front-end contribution page.
+   *
+   * @param int $contributionId
+   *   is the invoice contribution id (got created after processing participant payment).
+   * @param array $params
+   *   to take user provided input of transaction details.
+   * @param string $paymentType
+   *   'owed' for purpose of recording partial payments, 'refund' for purpose of recording refund payments.
+   * @param int $participantId
+   *
+   * @return array
+   */
+  public static function processAdditionalPayment($contributionId, $params, $paymentType = 'owed', $participantId = NULL) {
```

Comment:
I was looking at this & the payment.create api - which should be basically the same function - and there doesn't seem to be much overlap!

Reply:
Since this is the same path that we follow while doing partial payment from the back end forms, this PR also follows the same route for the front end payments too. I've looked into payment create api and wasn't able to see any param that actually triggers the correct receipt to the users.

Reply:
@monishdeb could you suggest to @jitendrapurohit the appropriate calls so there is good code reuse?

Reply:
@eileenmcnaughton @JoeMurray @jitendrapurohit Instead of having a separate function we can call CRM_Contribute_BAO_Contribution::addPayments(..) inside recordAdditionalPayment(..), ONLY for partial payment. But then addPayments(..) itself is only responsible to create proportional entries. Here's my [patch](https://gist.github.com/monishdeb/8dc327fb3f7765729d554630cfc72528) to does this change. You can then call this same function for online form too.

On the other hand, in order to extend Payment.create API to handle partial payment and even refund, we must have payment details in the ```$params``` and eventually it will call processAdditionalPayment(...) to handle additional owed or refund amount, the intent is to move the code towards BAO fn from API. What do you think?

Reply:
@monishdeb right - it seems Payment.create api needs to accept more parameters. The goal is the code logic should not be in that function - but that should be the function we call instead of a BAO function from any form.